File: backend/server/src/main/kotlin/nl/knaw/huc/dexter/api/Corpus.kt
package nl.knaw.huc.dexter.api

import java.time.LocalDate
import java.time.LocalDateTime
import java.util.*

data class FormCorpus(
    val title: String,
    val description: String? = null,
    val rights: String? = null,
    val access: AccessType? = null,
    val parentId: UUID? = null,
    val location: String? = null,
    val earliest: LocalDate? = null,
    val latest: LocalDate? = null,
    val contributor: String? = null,
    val notes: String? = null,
    val ethics: String? = null
)

data class ResultCorpus(
    val id: UUID,
    val parentId: UUID?,
    val title: String,
    val description: String? = null,
    val rights: String? = null,
    val access: AccessType? = null,
    val location: String? = null,
    val earliest: LocalDate? = null,
    val latest: LocalDate? = null,
    val contributor: String? = null,
    val notes: String? = null,
    val ethics: String? = null,
    val createdBy: UUID,
    val createdAt: LocalDateTime,
    val updatedAt: LocalDateTime
)

/**
 * With associated child resources
 */
data class ResultCorpusWithChildIds(
    val id: UUID,
    val parentId: UUID?,
    val title: String,
    val description: String? = null,
    val rights: String? = null,
    val access: AccessType? = null,
    val location: String? = null,
    val earliest: LocalDate? = null,
    val latest: LocalDate? = null,
    val contributor: String? = null,
    val notes: String? = null,
    val ethics: String? = null,
    val createdBy: UUID,
    val createdAt: LocalDateTime,
    val updatedAt: LocalDateTime,

    val tags: List<Int>,
    val languages: List<String>,
    val sources: List<UUID>,
    val metadataValues: List<UUID>,
    val media: List<UUID>,
    val subcorpora: List<UUID>
)

fun ResultCorpus.toResultCorpusWithChildIds(
    tags: List<Int>,
    languages: List<String>,
    sources: List<UUID>,
    metadataValues: List<UUID>,
    media: List<UUID>,
    subcorpora: List<UUID>
) = ResultCorpusWithChildIds(
    id = id,
    parentId = parentId,
    title = title,
    description = description,
    rights = rights,
    access = access,
    location = location,
    earliest = earliest,
    latest = latest,
    contributor = contributor,
    notes = notes,
    ethics = ethics,
    createdBy = createdBy,
    createdAt = createdAt,
    updatedAt = updatedAt,

    tags = tags,
    languages = languages,
    sources = sources,
    metadataValues = metadataValues,
    media = media,
    subcorpora = subcorpora
)

File: frontend/src/model/Corpus.ts
import { Any } from '../components/common/Any';
import { Source } from './Source';
import { ResultTag } from './Tag';
import { MetadataValue } from './Metadata';
import { ResultLanguage } from './Language';
import { UUID, WithId } from './Id';
import { Access } from './Access';
import { LocalDate, LocalDateTime } from './Date';
import _ from 'lodash';

/**
 * Corpus form as required by server
 */
export type FormCorpus = {
  title: string;
  description?: string;
  rights?: string;
  access?: Access;
  parentId?: UUID;
  location?: string;
  earliest?: LocalDate;
  latest?: LocalDate;
  contributor?: string;
  notes?: string;
  ethics?: string;
};
/**
 * Corpus result as sent by server
 */
export type ResultCorpus = FormCorpus &
  WithId & {
    createdBy: UUID;
    createdAt: LocalDateTime;
    updatedAt: LocalDateTime;
  };
export type ResultCorpusWithChildIds = ResultCorpus & {
  tags: number[];
  languages: string[];
  sources: UUID[];
  metadataValues: UUID[];
  media: UUID[];
  subcorpora: UUID[];
};

export function toResultCorpusWithChildren(
  result: ResultCorpus,
): ResultCorpusWithChildIds {
  return {
    ...result,
    tags: [],
    languages: [],
    sources: [],
    metadataValues: [],
    media: [],
    subcorpora: [],
  };
}

/**
 * Corpus including child resources
 */
export type Corpus = Omit<ResultCorpus, 'parentId'> & {
  parent?: ResultCorpus;
  tags: ResultTag[];
  languages: ResultLanguage[];
  sources: Source[];
  metadataValues: MetadataValue[];
  subcorpora: Corpus[];
};

export function isCorpus(toTest: Any): toTest is Corpus {
  if (!toTest) {
    return false;
  }
  return (
    !_.isUndefined((toTest as Corpus)?.title) &&
    !_.isUndefined((toTest as Corpus)?.contributor)
  );
}

export type SubmitFormCorpus = Omit<Corpus, 'id'>;
